File: array/189/189.js
// solution 1
var rotate = function(nums, k) {
  let extra = [];
  for (let i = 0; i < nums.length; i++) {
    extra[(i + k) % nums.length] = nums[i];
  }
  for (let i = 0; i < nums.length; i++) {
    nums[i] = extra[i];
  }
};

// solution 2
var rotate = function(nums, k) {
  const reverse = (arr, start, end) => {
    while (start < end) {
      let temp = arr[start];
      arr[start] = arr[end];
      arr[end] = temp;
      start++;
      end--;
    }
  };
  // if we rotate an array equal to its size, it will turn out to be the same, so we only need to take effective steps.
  //if size=5, for k=3 and k=8, the array will be the same after rotation
  k = k % nums.length;
  reverse(nums, 0, nums.length - 1);
  reverse(nums, 0, k - 1);
  reverse(nums, k, nums.length - 1);
};

File: array/189/189.py
# solution 1
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        extra = [None]*len(nums)
        for i in range(len(nums)):
            extra[(i+k) % len(nums)] = nums[i]

        for i in range(len(nums)):
            nums[i] = extra[i]

#solution 2 
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        def reverse(arr,start,end):
            while start<end:
                temp=arr[start]
                arr[start]=arr[end]
                arr[end]=temp
                start,end=start+1,end-1
        k%=len(nums)
        reverse(nums,0,len(nums)-1)
        reverse(nums,0,k-1)
        reverse(nums,k,len(nums)-1)